### ISSUE TITLE
Create a bounded version

### ISSUE THREAD

**Issue by jfischoff** [jfischoff/postgresql-queue#5]

**Comment by steshaw** [comment]
Curious what you mean by this? Do you mean to bound the "depth" of the queue? i.e. the number of rows that can appear `enqueued` in the payload table?

**Comment by jfischoff** [comment]
>  Do you mean to bound the "depth" of the queue? i.e. the number of rows that can appear enqueued in the payload table?

Yes exactly. I would like to have enqueueing fail if the queue is too backed up. That way I can propagate then back pressure to the consumer.

**Comment by steshaw** [comment]
I can see how that would be helpful in various use cases. In my current use case, we're wanting to delay the sending of emails via the queue. These are critical to our users (hence the persistent queue) and so I'd want to just keep putting them on the queue (until we run out of disk space in the db, I guess...). If it were a bounded queue, the upstream user would only be able to shed those precious emails. So I hope it's a configuration option :).

I'm really glad you started this project because I didn't know how to get started with LISTEN/NOTIFY from Haskell PostgreSQL libraries. On my [retries branch](https://github.com/steshaw/postgresql-queue/tree/retries), I've made a number of possibly dubious and non-configurable "options" to suit our use-case for now. It's also a work in progress. I'd be happy to contribute those changes in a nicer configurable way if those are the kinds of things you'd like to incorporate. For now, it was nice to be able to hack it to do want we wanted. Thanks!

**Comment by jfischoff** [comment]
> If it were a bounded queue, the upstream user would only be able to shed those precious emails. So I hope it's a configuration option :).

I think it would be a different module with a similar interface. The bounded version is going to be _much_ slower I imagine and like you have pointed out it is not needed in all situations.

> I'm really glad you started this project because I didn't know how to get started with LISTEN/NOTIFY from Haskell PostgreSQL libraries. 

Thanks. That was one of me reasons for starting the project (to play with LISTEN/NOTIFY) and I'm glad I saved you the hassle of learning how to use them :).

> On my retries branch, I've made a number of possibly dubious and non-configurable "options" to suit our use-case for now

I'll definitely take a look!
